util/progress/progressui/display.go

```diff
@@ -265,6 +274,140 @@ func (d *plainDisplay) done() {
 	d.t.printErrorLogs(d.t.w)
 }
 
+type jsonDisplay struct {
+	t    *trace
+	logs *logMux
+	enc  *json.Encoder
+}
+
+// newJSONDisplay creates a new SolveStatusDisplay that outputs the completed vertex
+// progress indicator for each vertex in JSON format.
+func newJSONDisplay(w io.Writer, opts ...DisplayOpt) Display {
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", "  ")
+	return Display{
+		disp: &jsonDisplay{
+			t:    newTrace(w, false),
+			logs: newLogMux(),
+			enc:  enc,
+		},
+	}
+}
+
+func (d *jsonDisplay) init(displayLimiter *rate.Limiter) {
+	// Display limiter is ignored for the JSON display.
+}
+
+func (d *jsonDisplay) update(ss *client.SolveStatus) {
+	d.t.update(ss, 80)
+
+	// JSON format keeps the logs separated into their streams
+	// and takes the aggregated result. The trace does not do that
+	// so we use and update logMux for log messages by digest/stream.
+	for _, l := range ss.Logs {
+		d.logs.update(l)
+	}
+
+	// Only look at the digests that have updates.
+	// If a digest doesn't have an update, it can't
+	// have been completed.
+	for dgst := range d.t.updates {
+		v, ok := d.t.byDigest[dgst]
+		if !ok {
+			continue
+		}
+
+		if v.isCompleted() {
+			d.printVtx(v)
+		}
+		delete(d.t.updates, dgst)
+	}
+}
+
+func (d *jsonDisplay) printVtx(v *vertex) {
+	intervals := make([]VertexInterval, len(v.mergedIntervals))
+	for i, interval := range v.mergedIntervals {
```

Comment:
I think this prints same vertex multiple times with all intervals. The logic should be to print interval after it completes and if there is another interval later then it prints a new object but without the previous data.

Reply:
I presently have it so the output gets printed when the entire vertex completes which means we have the completed aggregate data from all of the intervals. Were we going to print a JSON object for each interval that completes?

Reply:
The vertex does complete, but you don't know if there will be another run for the same vertex (eg. some runs needs to happen for cache lookup and later for execution, but there are more complicated cases).

Reply:
Fixed this. I didn't realize that the intervals were caused by a vertex completing and then starting again. We're going to print the json object when the vertex completes, erase any state associated with it, and print a new one if it runs again. Most builds this shouldn't happen with but there are some cases where a vertex can run multiple times.